[ios][sdk-54][precompile] added support for USE_FRAMEWORKS with Expo modules

chrfalch opened the pull request:
# Why

Some Expo modules are including React Native Core header files to override base RN classes in their public headers. This causes the swift interop generation to fail, since the RN headers are not fully modular - some include c++ code which fails when generating the module interface for these modules.

This isn't a problem for the module itself - but for the modules or targets that are using them. XCode will try to generate a public and modular interface for the module using the Objective-c headers - but when these contains c++ the process fails.

**NOTE:** Until we get a truly modular set of headers for React Native using USE_FRAMEWORKS will fail with React Native precompiled binaries. Building from source or building without USE_FRAMEWORKS does not fail since header files are detected through search paths only - without any checks for public/modularity.

This PR contains a workaround and not a final fix that we can implement when the above is finished (it is in the works!).

Closes #39080

# How

This commit fixes this by disabling frameworks (forcing static linking) of the targets that uses RN headers in their own public headers (Expo, ExpoModulesCore, ReactAppDependencyProvider, expo-dev-menu).

To fix this I added an installer.rb class to expo-modules-autolinking that overrides the Podfile and Installer and patches the necessary modules. Also added `additionalFrameworkModulesToPatch` parameter to use_expo_modules to be able to add additional framework modules to patch when using USE_FRAMEWORKS. Not sure if this is needed, but it will be very nice to have if someone needs it one day!

edit: for clarification, this makes sure a few of our own core modules (listed above) are being statically linked, while leaving all others as frameworks - it does not switch to build from source.

# Test Plan

Tested with new Expo app using SDK-54 and the changes from this commit:

✅ Build without precompiled RN, no USE_FRAMEWORKS
✅ Build without precompiled RN, USE_FRAMEWORKS=static 
✅ Build without precompiled RN, USE_FRAMEWORKS=dynamic 
✅ Build with precompiled RN, no USE_FRAMEWORKS
✅ Build with precompiled RN, USE_FRAMEWORKS=static
✅ Build with precompiled RN, USE_FRAMEWORKS=dynamic

# Checklist

- [x] I added a `changelog.md` entry and rebuilt the package sources according to [this short guide](https://github.com/expo/expo/blob/main/CONTRIBUTING.md#-before-submitting)
- [x] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

chrfalch committed:
[ios][sdk-54][precompile] added support for USE_FRAMEWORKS with Expo

Some Expo modules are including React Native Core header files to override base RN classes in their public headers. This causes the swift interop generation to fail, since the RN headers are not fully modular - some include c++ code which fails when generating the module interface for these modules.

This isn't a problem for the module itself - but for the modules or targets that are using them. XCode will try to generate a public and modular interface for the module using the Objective-c headers - but when these contains c++ the process fails.

Untill we get a truly modular set of headers for React Native using USE_FRAMEWORKS will fail with React Native precompiled binaries. Building from source or building without USE_FRAMEWORKS does not fail since header files are detected through search paths only - without any checks for public/modularity.

This commit fixes this by disabling frameworks (forcing static linking) of the targets that uses RN headers in their own public headers (Expo, ExpoModulesCore, ReactAppDependencyProvider, expo-dev-menu).

To fix this we are inserting a pre_install hook in the pod file that fixes the issues with react-core-prebuilt:

- Added pre_install script in expo/scripts
- Added call to pre_install in Podfiles:
  - Bare-Expo/Podfile
  - expo-template-bare-minimum/ios/Podfile
  - native-tests/ios
  - paper-tester/ios (not really necessary since this is only a problem on the new arch).

Tested in mulitple scenarios:

With new Expo app using SDK-54 and the changes from this commit:

✅ Build without precompiled, no USE_FRAMEWORKS
✅ Build without precompiled, USE_FRAMEWORKS=static
✅ Build without precompiled, USE_FRAMEWORKS=dynamic
✅ Build with precompiled, no USE_FRAMEWORKS
✅ Build with precompiled, USE_FRAMEWORKS=static
✅ Build with precompiled, USE_FRAMEWORKS=dynamic

chrfalch committed:
Update packages/expo/CHANGELOG.md

Co-authored-by: Expo Bot <34669131+expo-bot@users.noreply.github.com>

Kudo commented on apps/bare-expo/ios/Podfile:
```diff
@@ -24,6 +24,10 @@ abstract_target 'BareExpoMain' do
 
+  pre_install do |installer|
+    expo_pre_install!(installer)
+  end
```
requiring the template change is not ideal, especially that's a temporary workflow. we had a cocoapods monkey patching code in https://github.com/expo/expo/tree/main/packages/expo-modules-autolinking/scripts/ios/cocoapods
any change to do the pre_install hook in the hook?

chrfalch replied:
I've now implemented the logic in a class called `installer.rb` as you suggested - this seems to work fine so I've remove the changes to our pod files and templates! Great suggestion!

chrfalch committed:
[ios][sdk-54] added patching of installer/podfile to run pre_install

Instead of us having to add pre_install hook to all of the podfiles and templates, we've now added a class that basically sublcasses the Podfile and installer to patch the framework code (thanks @kudo).

- Removed template changes
- Removed podfile changes
- Removed react-core-prebuilt.rb
- Added installer.rb that overrides and patches modules if necessary
- Added extra option to use_expo_modules (`additionalFrameworkModulesToPatch`) that let's you specify other modules to include when patching

chrfalch committed:
[ios][pod] added comment

Added comment explaining what the installer overrides is actually doing!

chrfalch committed:
Merge branch 'main' into chrfalch/sdk-54-use_frameworks_ios_fix

Kudo commented on packages/expo-modules-autolinking/scripts/ios/cocoapods/installer.rb:
```diff
@@ -0,0 +1,76 @@
+    # Returns the linkage type if USE_FRAMEWORKS is set, otherwise returns nil
+    def get_linkage?()
+      props = JSON.parse(File.read(File.join(installer.project_root, 'Podfile.properties.json'))) rescue {}
+      linkage_str = (ENV["USE_FRAMEWORKS"] || props['ios.useFrameworks'] || nil)
```
ditto to remove `ios.useFrameworks` check. react-native also set the env from https://github.com/facebook/react-native/blob/226d2ba1607fe34f8dc16616e7ee7182a2d7a309/packages/react-native/scripts/cocoapods/utils.rb#L245-L261

if the env from react-native is too late to set, we can try to use `target_definition.build_type.to_s` to get the linkage type from cocoapods

chrfalch replied:
Removed pod file props from `should_disable_use_frameworks_for_core_expo_pods?` as well, these are not needed since all of the required env variables were already set in the pre_install step. Nice catch as always, @Kudo!

Kudo commented on packages/expo-modules-autolinking/scripts/ios/cocoapods/installer.rb:
```diff
@@ -0,0 +1,76 @@
+
+    def expo_add_modules_to_patch(modules)
+      framework_modules_to_patch.concat(modules)
+      Pod::UI.puts "[Expo] ".blue + "Added extra framework modules to patch: #{modules.join(', ')}"
```
```suggestion
```

the following log will show all `framework_modules_to_patch` which is good enough

Kudo commented on packages/expo-modules-autolinking/scripts/ios/cocoapods/installer.rb:
```diff
@@ -0,0 +1,76 @@
+      nil
+    end
+  end
+end
```
add empty line at EOF

chrfalch committed:
Update packages/expo-modules-autolinking/scripts/ios/cocoapods/installer.rb

Co-authored-by: Kudo Chien <kudo@expo.dev>

chrfalch committed:
codereview: added extra line at en

chrfalch committed:
Merge branch 'chrfalch/sdk-54-use_frameworks_ios_fix' of github.com:expo/expo into chrfalch/sdk-54-use_frameworks_ios_fix

chrfalch committed:
codereview: only use ENV for linkage detection

React native already sets the USE_FRAMEWORKS env variable, so there is no need for us to check the pod props file.
